### PR TITLE
Chat: fold tool-call runs, Tasks row for Recent chats, peek header in topbar

### PR DESCRIPTION
**Pull request by AkshilVT**
## Summary

Native chat (`native_chat_enabled`) only; the legacy `templates/claude` iframe is untouched.

**Tool-call runs fold into one chip.** Two or more back-to-back tool calls (Bash, Read, Edit, …) with nothing between them render as a single `N tool calls` accordion once every call in the stretch has settled. While any call is still running the chips stay individual, so progress is visible. Clicking the run header expands it into the N individual tool chips, each its own collapsed accordion with its own persisted open state. A text/thinking segment, or a permission card filed under a chip, breaks the run. Pure `groupToolRuns()` in `protocol/segments.ts`, indices preserved for `cardKey` / `cardsAfter` / the streaming typer.

**Recent chats reuse the Tasks list row.** The chat home's Recent panel now draws the Tasks page's own list-view item (`TaskNode`, chat variant via `TaskRowItem`): status ring, TASK-nnn, title (honours `task_card_last_message`), file mark, message count, time. Same ordering as the Tasks list (`sortForList`: lanes → drafts first → time). Stripped for the chat: no caret slot, no Open button, no folder chip, no pin/peek. Row press opens the chat in place and marks it read. Rows come from `/api/tasks` filtered to the pane's folder (or file); `RecentRow` / `useRecentSessions` are gone.

**Chat header reuses the side-peek identity line.** When the open chat has a task row, the topbar shows the peek header's ring · TASK-nnn · title · project chip (`TaskPeekWho` / `TaskPeekProject`, extracted verbatim from `TaskPeek`). Brand-new chats fall back to the old ✻ Claude line.

`check-boundaries.mjs` gains a narrow per-module allowlist so `apps/claude` may import `shell/ScheduleTaskViews`, `shell/tasks-lib`, `shell/TaskPeekWho` (lifting them to `platform/` would drag several thousand lines of tasks-lib along).

## Test plan

- [x] `tsc --noEmit`, `check-boundaries`, `bun test src` (5898 pass), `pytest` template pins (29 pass)
- [x] cmux browser QA: fold/expand/collapse/persist on two settled transcripts; Recent rows geometry/order/pref toggle vs /tasks; header identity vs /tasks side peek; /tasks list + peek unchanged; zero console errors
- [ ] Manual: send a multi-tool prompt, watch chips stay individual while running and fold on settle

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches landing data source, shell import boundaries, and transcript grouping with live-poll edge cases; broad test coverage but behavior spans chat boot, navigation, and unread clearing.
> 
> **Overview**
> **Recent chats** now load from **`/api/tasks`** (not agent `sessions`), render with the Tasks page **`TaskRowItem`** row, and share **`sortForList`** / pane filtering via **`useRecentTasks`**. **`RecentRow`** and **`useRecentSessions`** are removed; **`check-boundaries`** whitelists three shell modules so the chat can reuse task row, sort, and peek UI without copying them.
> 
> **Chat chrome:** the topbar shows **`TaskPeekWho`** when a task row exists (skeleton while **`useSessionTask`** is pending; **`seedSessionTask`** on list press avoids wrong identity on deep links). **Never-sent chat draft** rows call **`onOpenChatDraft`**; **`Composer`** gets **`focusRequest`** / post-draft caret focus so explorer **`noFocus`** hosts still focus the box.
> 
> **Transcript:** consecutive settled tool calls fold into **`ToolRunChip`** runs via **`groupToolRuns`** (live trailing stretch stays unfolded; cards break runs); **`SegmentView`** / **`Turn`** pass **`live`**.
> 
> Tests and boot mocks updated for **`/api/tasks`** and the new row/header/focus behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 50dab7ab1ed53ac3cd9149ea533af6d4fa2e2076. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: frontend/src/apps/claude/protocol/segments.ts [hunk 1]
```diff
@@ -48,6 +48,134 @@ export function cardKey(seq: number, seg: Segment | undefined | null, i: number)
...
+    let from = i;
+    for (let k = i; k < end; k += 1) {
+      if (!cardsAfter || !cardsAfter.has(k)) continue;
+      push(from, k);
+      out.push(list[k] as Segment);
+      from = k + 1;
+    }
+    push(from, end);
+    i = end;
+  }
```
[the remark this answers is not in the document]
Comment (AkshilVT) [reply frontend/src/apps/claude/protocol/segments.ts:194]: Fixed in e1dd204. `groupToolRuns` now takes the turn's `streaming` flag: while it is on, the stretch that runs to the END of the segment list is held open, because that is the only one the poll can still append to. A stretch with anything after it is closed and still folds the moment it settles, so nothing mid-transcript waits. `Turn` passes `live={!!turn.streaming}` through `SegmentView`. Done still equals combined — the tail folds once, at turn end. Covered by two new cases in `protocol/segments.test.ts` and one in `ui/tool-run.test.tsx`.

File: frontend/src/apps/claude/ui/useRecentTasks.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+   *  the same fact): the identity of the subscription function is not a fact
+   *  about the session, so a caller that passes a fresh closure — every test
+   *  does — must not re-subscribe this hook on every render. Read at subscribe
+   *  time, which is the only moment it is used. */
+  const subscribeRef = useRef(subscribe);
+  subscribeRef.current = subscribe;
+  useEffect(() => {
+    if (!sessionId) return;
+    return subscribeRef.current(file, (next) => setRows(next));
+  }, [sessionId, file]);
```
[the remark this answers is not in the document]
Comment (AkshilVT) [reply frontend/src/apps/claude/ui/useRecentTasks.ts:341]: Fixed in e1dd204. `useSessionTask` now keeps the same `painted` ref `useRecentTasks` does: the opening `null` is honoured only before the first real answer of the hook's life, and swallowed after, so a re-entry's re-subscribe no longer blanks the row the topbar is already naming. New case in `ui/useRecentTasks.test.tsx`.

File: frontend/src/apps/claude/ClaudeChat.tsx [hunk 3]
```diff
@@ -2132,6 +2133,28 @@ function ChatBody(props: ChatBodyProps) {
...
+   *
+   * NO `controller.openSession`, and no `newChat()` either: the landing's
+   * controller is already on a fresh chat, and re-cutting it here would throw
+   * away a run this page may have adopted in the background.
+   */
+  const onOpenChatDraft = useCallback(() => {
+    resetCardPolicy(cardPolicy);
+    setEntered(true);
+    setStranded(null);
+  }, [cardPolicy]);
```
[the remark this answers is not in the document]
Comment (AkshilVT) [reply frontend/src/apps/claude/ClaudeChat.tsx:2183]: Fixed — the transport now closes the window rather than the press. `platform/lib/drafts.ts` keeps an `inflightChat` map of chat writes still on the wire, one per key, and `fetchChatDraft` awaits the entry for its key before it reads. React runs the deleted subtree's effect cleanups before the new subtree's effects, so the landing composer's unmount flush is always DISPATCHED first; what was missing was only that the new mount's seed could be answered out of a snapshot taken before that PUT landed. Same shape as the existing `spent` set, and at module scope for the same reason it gives — the flush belongs to a component being thrown away and the seed runs in a different one. Two cases in `drafts.test.ts`; the first fails on the old code with "the words as they were one draft ago".

File: frontend/src/platform/lib/drafts.ts [hunk 4]
```diff
@@ -466,6 +515,12 @@ export async function fetchDrafts(): Promise<DraftsSnapshot | null> {
  *  start of the wait or in the middle of it makes no difference to the reader:
  *  the words came back after they were sent. */
 export async function fetchChatDraft(key: string): Promise<ChatDraft | null> {
+  if (spent.has(key)) return null;
+  // A WRITE ON THIS KEY STILL IN THE AIR IS WAITED FOR FIRST (see
+  // `inflightChat`): the composer that is being replaced flushes on unmount,
+  // and reading past that PUT hands the new box the previous draft.
+  const pending = inflightChat.get(key);
+  if (pending) await pending;
```
[the remark this answers is not in the document]
Comment (AkshilVT) [reply frontend/src/platform/lib/drafts.ts]: Correct, and fixed in 51d7d84. `inflightChat` now holds a SET per key and `fetchChatDraft` awaits all of it, so the GET is not dispatched until every write that was already on the wire has landed. What is awaited is a snapshot taken when the read starts — a write dispatched after that is genuinely concurrent with the GET, and waiting on those too would be a wait with no end under fast typing. The new case asserts the contract (nothing is GET-ed until both PUTs have landed) rather than the returned string, since a string comparison here passes or fails on microtask ordering; it fails against the one-slot version with `["the flushed words"]`.

File: frontend/src/apps/claude/ClaudeChat.tsx [hunk 5]
```diff
@@ -1784,7 +1785,11 @@ function ChatBody(props: ChatBodyProps) {
    * declared with the other gestures further down.
    */
   const [leftLive, setLeftLive] = useState(false);
-  const recent = useRecentSessions(
+  /** Bumped by a gesture that has asked for the composer — today only the
+   *  never-sent chat's row (`onOpenChatDraft`). Handed to the CHAT's composer
+   *  alone: the landing's box is the one the reader is leaving. */
+  const [focusReq, setFocusReq] = useState(0);
```
[the remark this answers is not in the document]
Comment (AkshilVT) [reply frontend/src/apps/claude/ClaudeChat.tsx:1791]: Fixed in 50dab7a. The request is now SPENT: `setFocusReq(0)` on Back — the funnel out of the chat — and again in `onOpenSession`, beside the `setStranded(null)` that is already spelled in both for the same reason. So a press on a never-sent row asks for the caret once, and the next ordinary session opened under `noFocus` leaves the listing its keyboard. Note on coverage: the consuming half is tested (Composer.test.tsx, "a GESTURE that asked for the box beats a host that keeps the keyboard"); the host-side reset is not covered by a new test — driving it needs a full ClaudeChat mount with the task listing stubbed, a draft press, a Back and a second open, which is a large and flake-prone harness for a two-line state clear. Flagging that rather than implying coverage.
